sycl/include/CL/sycl/detail/spirv.hpp

```diff
@@ -69,6 +70,221 @@ T GroupBroadcast(T x, id<Dimensions> local_id) {
   return __spirv_GroupBroadcast(group_scope<Group>::value, ocl_x, ocl_id);
 }
 
+// Single happens-before means semantics should always apply to all spaces
+static inline constexpr __spv::MemorySemanticsMask::Flag
+getMemorySemanticsMask(intel::memory_order order) {
+  __spv::MemorySemanticsMask::Flag spv_order = __spv::MemorySemanticsMask::None;
+  switch (order) {
+  case intel::memory_order::relaxed:
+    spv_order = __spv::MemorySemanticsMask::None;
+  case intel::memory_order::__consume_unsupported:
+  case intel::memory_order::acquire:
+    spv_order = __spv::MemorySemanticsMask::Acquire;
+  case intel::memory_order::release:
+    spv_order = __spv::MemorySemanticsMask::Release;
+  case intel::memory_order::acq_rel:
+    spv_order = __spv::MemorySemanticsMask::AcquireRelease;
+  case intel::memory_order::seq_cst:
+    spv_order = __spv::MemorySemanticsMask::SequentiallyConsistent;
+  }
+  return static_cast<__spv::MemorySemanticsMask::Flag>(
+      spv_order | __spv::MemorySemanticsMask::SubgroupMemory |
+      __spv::MemorySemanticsMask::WorkgroupMemory |
+      __spv::MemorySemanticsMask::CrossWorkgroupMemory);
+}
+
+static inline constexpr __spv::Scope::Flag getScope(intel::memory_scope scope) {
+  switch (scope) {
+  case intel::memory_scope::work_item:
+    return __spv::Scope::Invocation;
+  case intel::memory_scope::sub_group:
+    return __spv::Scope::Subgroup;
+  case intel::memory_scope::work_group:
+    return __spv::Scope::Workgroup;
+  case intel::memory_scope::device:
+    return __spv::Scope::Device;
+  case intel::memory_scope::system:
+    return __spv::Scope::CrossDevice;
+  }
+}
+
+template <typename T, access::address_space AddressSpace>
+inline typename detail::enable_if_t<std::is_integral<T>::value, T>
+AtomicCompareExchange(multi_ptr<T, AddressSpace> mptr,
```

Comment:
I am not sure, but probably we need to  disable all Atomic* templates for constant and private address space?

Reply:
You're right that those address spaces aren't supported, and the user-facing API (`atomic_ref`) performs this check.  Do you think we need to add the checks to the implementation as well?  These functions should only be called inside `atomic_ref`.

Reply:
I see, check in the user-facing API looks enough, thanks!